---
layout: news
title: What's new in TSKIT 0.6.0!
description: The 0.6.0 release of tskit is a major release with significant enhancements to genetic analysis capabilities, tree visualization, and general usability. 
timestamp: 2024-11-05
author: Ben Jeffery
---

The 0.6.0 release of tskit is a major release with significant enhancements to genetic
analysis capabilities, tree visualization, and general usability. 

### Breaking changes

The definition of `TreeSequence.genetic_relatedness` and
`TreeSequence.genetic_relatedness_weighted` are changed to *average* over sample sets.
This allows comparison between sample sets of different sizes. For example this will change the result
by a factor of four for diploid sample sets. The default for these methods is also changed to
`polarised=True`, but the output is unchanged for `centre=True` (the default). Thanks to 
[@petrelharp](https://github.com/petrelharp) and [@mmosmond](https://github.com/mmosmond) for their work on this.

### New features
Thanks to [@nspope](https://github.com/nspope), `tskit` now has a number of new methods for
calculating coalescence events and rates. The `pair_coalescence_counts` method calculates
coalescence events per node or time interval, `pair_coalescence_quantiles` estimates quantiles
of pair coalescence times using empirical CDF inversion, and `pair_coalescence_rates` estimates
instantaneous rates of pair coalescence within time intervals from the empirical CDF.
These methods are useful for understanding the timings and patterns of shared ancestry.

```python
# Define time intervals ending with infinity over which to measure pair coalescence rates
>>> max_time = ts.tables.nodes.time.max()
>>> time_intervals = np.linspace(0, max_time, 50)
>>> time_intervals = np.append(time_intervals, np.inf)

# Compute pair coalescence rates
>>> ts.pair_coalescence_rates(time_intervals)
array([2.59651787e-04, 4.96541129e-05, 5.91296805e-05, 5.37257360e-05,
       4.65699674e-05, 4.56612744e-05, 3.94309857e-05, 4.96831488e-05,
       4.94545336e-05, 5.01056926e-05, 5.25042462e-05, 5.25537314e-05 ...
```

[@petrelharp](https://github.com/petrelharp), [@hfr1tz3](https://github.com/hfr1tz3), [@nspope](https://github.com/nspope), and [@avabamf](https://github.com/avabamf) have put together a new method `ts.extend_haplotypes` that reduces the number of edges in a tree sequence by extending ancestral haplotypes. This makes the tree sequence smaller, while adding unary nodes.

```python
>>> ts = msprime.sim_ancestry(10000, recombination_rate=0.001, sequence_length=10000)
>>> ts.num_edges
41512
>>> ts.extend_haplotypes().num_edges
40761
```


Several new methods have been added to the `Tree` class. [@Billyzhang1229](https://github.com/Billyzhang1229) has added both a `distance_between` method that calculates the total distance between two nodes in a tree, essentially adding  the branch lengths separating them, and an `rf_distance` method that calculates the unweighted Robinson-Foulds distance between two trees. [@hyanwong](https://github.com/hyanwong) has added an `ancestors` method that returns the ancestors of a given node.

```python
# Calculate the total distance between two nodes
>>> ts.at(547630.5).distance_between(42,43)
5578.6766720892065

# Calculate the Robinson-Foulds distance between two trees
>>> ts.at(547630.5).rf_distance(ts.at(647630))
192

# Get the ancestors of a node as an iterator
>>> list(ts.at(547630.5).ancestors(42))
[606, 931, 1863, 2483, 2570, 2590, 3603, 3913]
```

Several quality-of-life improvements have been added. Edges now have an `interval` attribute that returns a `tskit.Interval` object thanks to [@hyanwong](https://github.com/hyanwong) and those `Intervals` and `Trees` now have `mid` properties that return the midpoint of the interval or tree. Thanks to [@currocam](https://github.com/currocam) for this addition. Dropping metadata from a table is now simpler with the new `Table.drop_metadata` method, thanks to [@jeromekelleher](https://github.com/jeromekelleher).

```python
# Get the interval of an edge
>>> ts.edge(42).interval
Interval(left=194621.0, right=900640.0)

# Get the midpoint of an interval
>>> ts.edge(42).interval.mid
547630.5

# Drop metadata from a table
table.drop_metadata()
```

[@hyanwong](https://github.com/hyanwong) has also added a `states` method to `Variant` objects that returns the genotypes as an array of strings, rather than integer indexes, to aid comparison of genetic variation. Note that this is an inefficient way to access the genotypes, but can be useful for showing the actual alleleic states.

```python
>>> next(ts.variants()).states()
array(['G', 'G', 'G', 'C', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G',
       'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G', 'G',
       'G', 'C',], dtype='<U1')
```

[@jeromekelleher](https://github.com/jeromekelleher) and [@petrelharp](https://github.com/petrelharp) have added a `genetic_relatedness_matrix` method that computes the full pairwise genetic relatedness between and within pairs of sets of nodes from sample sets, and a `genetic_relatedness_vector` method that computes the product of the genetic relatedness matrix and a weight vector.

```python
>>> ts.genetic_relatedness_matrix(sample_sets=((0,1),(2,3)))
array([[ 2.37625e-05, -2.37625e-05],
       [-2.37625e-05,  2.37625e-05]])
```

Finally, the `draw_svg` methods now have a number of new features. Thanks to [@hyanwong](https://github.com/hyanwong). Gene annotations can now be added to the x-axis, titles to the top of figures and node and mutation titles can be added to the tree. In addition, the `order` parameter can be used to pass a subset of nodes to the `Tree.draw_svg` method, allowing subtrees to be visually collapsed. The `pack_untracked_polytomies` option allows large polytomies involving untracked samples to be summarised as a dotted line.